### PR TITLE
ROX-18482: Introduce interfaces for log interception

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR is a prefactor in a series of PRs to establish log interception of structured logs to create notifications.

The first step is to introduce options to the logger created via `LoggerForModule` as well as two interfaces abstracting the converting of log entries to a `storage.Notification`, as well as a loosely coupled interface to write notifications.

The follow-up PRs will implement the interfaces.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

N/A

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #7257** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7257" target="_blank">Graphite</a> 
    * **PR #7423** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7423" target="_blank">Graphite</a>  👈
      * **PR #7424** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7424" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/7423?utm_source=stack-comment).

File: pkg/logging/logger.go [hunk 1]
```diff
@@ -193,3 +200,14 @@ func (l *LoggerImpl) Debugf(template string, args ...interface{}) {
...
 	l.InnerLogger.Debugw(msg, keysAndValues...)
 }
+
+func (l *LoggerImpl) processStructuredLog(msg string, keysAndValues ...interface{}) {
+	// Short-circuit if no event writer or converter is found.
+	if l.opts.notificationWriter == nil || l.opts.notificationConverter == nil {
+		return
+	}
+
+	// We will use the log converter to convert logs to a storage.Event.
```
Comment (stehessel) [inline pkg/logging/logger.go:210]:
```suggestion
	// We will use the log converter to convert logs to a storage.Notification.
```

File: pkg/logging/logger.go [hunk 2]
```diff
@@ -193,3 +200,14 @@ func (l *LoggerImpl) Debugf(template string, args ...interface{}) {
 func (l *LoggerImpl) Debugw(msg string, keysAndValues ...interface{}) {
 	l.InnerLogger.Debugw(msg, keysAndValues...)
 }
+
+func (l *LoggerImpl) processStructuredLog(msg string, keysAndValues ...interface{}) {
```
Comment (stehessel) [inline pkg/logging/logger.go:204]: Is this function intended to do something other than create notifications? If not maybe a more concrete name like `createNotification` or `notifyFromLog` or something like that would be better suited.

File: pkg/notifications/writer.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package notifications
+
+import (
+	"github.com/stackrox/rox/generated/storage"
+)
+
+// Writer is capable of writing notifications. The notification may be written to an underlying storage.
+type Writer interface {
```
Comment (stehessel) [inline pkg/notifications/writer.go:8]: I guess this is no longer needed with https://github.com/stackrox/stackrox/pull/7509?

File: compliance/collection/main.go [hunk 4]
```diff
@@ -5,7 +5,7 @@ import (
 
 	"github.com/stackrox/rox/compliance/collection/compliance"
 	"github.com/stackrox/rox/pkg/env"
-	"github.com/stackrox/rox/pkg/retry"
+	"github.com/stackrox/rox/pkg/retry/handler"
```
Comment (stehessel) [inline compliance/collection/main.go:8]: Why this change?

Comment (dhaus67) [reply compliance/collection/main.go:8]: Since we want to make use of `retry.WithRetry` we import the package in `pkg/logging`.

The `Unconfirmed message handler` that was previously defined within `retry` imports `pkg/logging` however.

Split this out into the different `handler` package so that `pkg/logging` can safely import it without trouble (i.e. cyclic imports). By the looks of it, not sure whether it should ever have been a part of the `retry` package.

**Review by stehessel — APPROVED** [review]
